core/src/main/java/org/infinispan/statetransfer/CommitManager.java

```diff
@@ -123,17 +125,22 @@ public final void commit(final CacheEntry entry, final Flag operation,
       });
    }
 
-   private void commitEntry(CacheEntry entry, InvocationContext ctx) {
+   private void commitEntry(CacheEntry entry, int segment, InvocationContext ctx) {
       if (!entry.isEvicted() && !entry.isRemoved() && entry.getValue() instanceof MergeOnStore) {
-         PersistenceUtil.loadAndComputeInDataContainer(dataContainer, persistenceManager, entry.getKey(), ctx, timeService, (k, oldEntry, factory) -> {
+         PersistenceUtil.loadAndComputeInDataContainer(dataContainer, segment, persistenceManager, entry.getKey(), ctx,
+               timeService, (k, oldEntry, factory) -> {
             Object newValue = ((MergeOnStore) entry.getValue()).merge(oldEntry == null ? null : oldEntry.getValue());
             if (newValue == null) {
                return null;
             }
             return factory.create(k, newValue, entry.getMetadata());
          });
       } else {
-         entry.commit(dataContainer);
+         if (segment != SegmentSpecificCommand.UNKNOWN_SEGMENT && entry instanceof ReadCommittedEntry) {
+            ((ReadCommittedEntry) entry).commit(segment, dataContainer);
+         } else {
+            entry.commit(dataContainer);
```

Comment:
We really need a `ContextEntry` interface where we can add these methods without worrying about exposing them via `AdvancedCache.getCacheEntry()`

Reply:
I agree, it is super messy :(